staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextensionsv1beta1client "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1beta1"
+)
+
+var (
+	// retries is the maximum number of attempts that a CRD will be given to establish
+	retries = 5
+	// interval is the amount of time that will pass between retries
+	interval = 500 * time.Millisecond
+	// factor is the scalar by which the prevous interval will be increased
+	factor = 1.0
+)
+
+// EnsureCRDEstablishment is a utility function that will ensure that a CustomResourceDefinition
+// is not only created, but ready for use by a Kubernetes cluster. If the CustomResourceDefinition
+// cannot be established within a reasonable amount of retries, an error will be returned.
+func EnsureCRDEstablishment(
+	i apiextensionsv1beta1client.ApiextensionsV1beta1Interface,
+	crd *apiextensionsv1beta1.CustomResourceDefinition,
+	stopCh <-chan struct{},
+) error {
+	// Attempt to create the crd (even if it's already present).
+	// If it is present, we still need to check its status.
+	if _, err := i.CustomResourceDefinitions().Create(crd); err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
So if the given crd is different, we will ignore that? Don't like it. Clearly, we can also not easily change the existing one. I tend to move the whole creation out of this func and just call it WaitForCRDEstablished.

Reply:
> So if the given crd is different, we will ignore that?

What do you mean? There are two possible cases here (CRD exists or not). If it does exist, than because of the `!apierrors.IsAlreadyExists(err)` check, the code will fall through to the same place as it would if it didn't exist and was created.

Reply:
So it does not really implement "ensure" because the existing crd might be outdated.

Reply:
by _outdated_ do you mean that the CRD is in a state where it would never be established? If so, than later login in this function would catch that case and return the appropriate error.

Reply:
No, I mean the case where a CRD already exists with that name, but the `crd` variable has a newer version.

Reply:
i.e. "ensure" sounds very much like "kubectl apply" + established.

Reply:
But the version field immutable. Hence, an error would occur when trying to create v2 if v1 already existed. For example, if you `kubectl apply` a v1 crd and try to apply a v2, you'll get the error:
```sh
The CustomResourceDefinition "foos.bar.io" is invalid: spec.version: Invalid value: "v2": field is immutable
```

Reply:
Hence, to bring back into context your question:
> So if the given crd is different, we will ignore that?

No, we would not ignore it. However, an error would be returned as is consistent with `kubectl apply` as the version field is immutable.

Reply:
I am using `Create` however and so I need to check if that will overwrite the existing one. I may need to check for existence and if already present, use `Patch`

Reply:
I think the easiest solution is just to turn the func into a simple WaitForEstablished helper. We cannot implement an apply-like logic easily.

Reply:
Are there apis from `kubectl` that we could leverage or would it still not be the easy.

Reply:
It won't be easy. Event the kubectl apply mechanism is known to be quirky. I think it's much better to leave that part to the implementor. He/she knows what to change. CRDs are too important to apply some half-baked heuristic to updates.

Reply:
Alright I'm thoroughly convinced 👍 Changes made.